src/plugins/memdump/userhook.cpp

```diff
@@ -719,6 +729,128 @@ void memdump::load_wanted_targets(const memdump_config* c)
     }
 }
 
+static event_response_t copy_on_write_ret_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto data = get_trap_params<memdump, copy_on_write_result_t<memdump>>(info);
+    memdump* plugin = data->plugin();
+    if (!data || !plugin)
+    {
+        PRINT_DEBUG("[MEMDUMP-USER] copy_on_write_ret_cb invalid trap params!\n");
+        drakvuf_remove_trap(drakvuf, info->trap, nullptr);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    if (!data->verify_result_call_params(info, drakvuf_get_current_thread(drakvuf, info)))
+        return VMI_EVENT_RESPONSE_NONE;
+
+    plugin->destroy_trap(drakvuf, info->trap);
+
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    page_info_t pinfo;
+    addr_t pa;
+
+    if (vmi_pagetable_lookup_extended(vmi, info->regs->cr3, data->vaddr, &pinfo) != VMI_SUCCESS)
+    {
+        PRINT_DEBUG("[MEMDUMP-USER] failed to get pa");
+        goto end;
+    }
+
+    pa = pinfo.paddr;
+
+    if (data->old_cow_pa == pa)
+    {
+        PRINT_DEBUG("[MEMDUMP-USER] PA after CoW remained the same, wtf? Nothing to do here...\n");
+        goto end;
+    }
+
+    PRINT_DEBUG("[MEMDUMP-USER] new PA %lx\n", pa);
+
+    for (auto& hook : data->hooks)
+    {
+        addr_t hook_va = ((data->vaddr >> 12) << 12) + (hook->trap->breakpoint.addr & 0xFFF);
+        PRINT_DEBUG("adding hook at %lx\n", hook_va);
+
+        make_trap(vmi, drakvuf, info, hook, hook_va);
+    }
+
+end:
+    drakvuf_release_vmi(drakvuf);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+static event_response_t copy_on_write_handler(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto plugin = get_trap_plugin<memdump>(info);
+    if (!plugin)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    addr_t vaddr = drakvuf_get_function_argument(drakvuf, info, 1);
+    addr_t pte = drakvuf_get_function_argument(drakvuf, info, 2);
+
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    page_info_t pinfo;
+    addr_t pa;
+    if (vmi_pagetable_lookup_extended(vmi, info->regs->cr3, vaddr, &pinfo) != VMI_SUCCESS)
```

Comment:
Again, just use `vmi_pagetable_lookup`

Reply:
The problem with `vmi_pagetable_lookup` is that it uses cache. As we deal with copy on write mechanism here, mappings between VA->PA are subject to change.

During our tests we encountered a bug which caused DRAKVUF incorrectly returning addresses from cache even though mappings were changed in the meantime. Replacing with `vmi_pagetable_lookup_extended` solves the problem, because it doesn't use cache.

Reply:
DRAKVUF periodically flushes the caches but if you have a usecase that's more sensitive to its frequency you can always just issue a `vmi_v2pcache_flush` in your plugin. You should not rely on `vmi_pagetable_lookup_extended` not using the cache, that's not something guaranteed by the API.